Support for all ZigBee frequencies

whoot opened the pull request:
Fix for #146

> Added support for all ZigBee frequencies (868, 784, 915 MHz). 
> Various bugfixes in tools
> Updated Readme

Commit:
bugfixes, added support for all frequencies

Committer: Jan Rude

Commit:
add missing kbutils

Commit:
delete dump

whoot committed:
Update zbfind

rmspeers commented on tools/zbdump:
```diff
@@ -46,8 +46,12 @@ parser.add_argument('-n', '--count', action='store', type=int, default=-1)
-if args.showdev:
-    show_dev()
+if args.showdev and not args.devstring:
+    print >>sys.stderr, "ERROR: Must specify an interface."
```
@whoot previously it was permitted for it to show all available interfaces in this case -- it shouldn't be just allowing you to do a -D if you provide a device string -- at least I always interpreted this behavior to be similar to zbid (and honestly am not sure if we really need showdev in most of these programs, as zbid should handle that for people).

whoot replied:
Oops, that is indeed a copy&paste mistake. I will correct that immediately.

rmspeers replied:
The -D without -i is again due to the usage of a Sewio.

whoot replied:
exactly. But why shouldn´t I fix problems that are Sewio specific in the Sewio branch?

rmspeers commented on tools/zbdump:
```diff
@@ -71,9 +75,13 @@ if not kb.is_valid_channel(args.channel):
+if args.channel == 0: rf_freq_mhz = 868
+elif args.channel >= 11 and args.channel <= 26: rf_freq_mhz = ((args.channel - 10) * 5 + 2400)
+elif args.channel >= 1 and args.channel <= 10: rf_freq_mhz = (args.channel * 2 + 904)
+elif args.channel >= 128 and args.channel <= 131: rf_freq_mhz = ((args.channel - 128) * 2 + 780)
```
@AdamLaurie I believe you instead were implementing different regions using pages, right? I presume that we should instead have this code follow your behavior, perhaps with adding page CLI options? Or, can we get it all lined up so that users always specify only a channel, and somewhere in the stack converts it to channel/page combinations?

I'm fine with keeping it all channel through the stack -- _but_ implementation needs to match between these changes and other areas of the code. I don't want two different code paths for frequency pages.

whoot replied:
Could you please be more specific about the 'pages'?
You can see in the documentation (https://www.sewio.net/open-sniffer/develop/http-rest-interface/) that the channel specifies the frequency.

I used a proxy to get verify the requests and every request goes to:
http://10.10.10.2/settings.cgi?chn=XX&modul=XX&rxsens=XX

rmspeers replied:
I'll let @AdamLaurie  share his thoughts and then we can figure out how to handle this.

AdamLaurie replied:
Thanks @rmspeers - the page thing is how it's done in Europe and/or on Silicon Labs hardware but it's not even consistent there... Some SiLabs firmware uses channels > 128 for the additional pages and some use chan/page. It may be the best way to deal with it is to move the "is_valid_channel" into the device driver code so each driver can make it's own mind up.

For SiLabs I have no choice - the page and channel are combined to create the correct config byte:

#subghz channels must be calculated to include page (3 LSB page + 5 LSB channel)
# e.g. page 28 channel 15 == '100' + '01111' = '10001111' = 0x8F

whoot replied:
I see. Thanks for clarification.
Actually I was planning to add the channels in the device driver code, but since it was only necessary in this tool, I added it there.

rmspeers commented on tools/zbfakebeacon:
```diff
@@ -49,7 +49,7 @@ if __name__ == '__main__':
     parser = argparse.ArgumentParser()
     parser.add_argument('-f', '--channel', '-c', action='store', dest='channel', required=True, type=int, default=11)
-    parser.add_argument('-i', '--interface', action='store', dest='devstring')
+    parser.add_argument('-i', '--interface', action='store', dest='devstring', required=True)
```
We have never had -i defined in tools as most interfaces are auto-discoverable if one is not provided, and thus users may not specify one if they are doing something simple.

whoot replied:
The -i parameter was already included.
I added the "required" argument, to get a correct error message when not using an interface. Otherwise the program quits with the error "KillerBee doesn´t understand device given by 'None'". Auto-discover does not work for me.

rmspeers commented on tools/zbfind:
```diff
@@ -1106,9 +1107,12 @@ class ZBPoller(Thread):
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
+    parser.add_argument('-i', '--iface', '--dev', action='store', dest='devstring', required=True)
```
similarly should not be required and default to None

whoot replied:
as already pointed out: without specifying an interface, the tool wont work (for me):
```
python zbfind -c 11

Error initializing KillerBee Sniffer
```

rmspeers replied:
Yes, this is due to a specific thing where you are only using the Sewio sniffer, which doesn't support auto-discovery at all due to the API/etc.

Although this change makes perfect sense in your case, we need to ensure that the behavior users expect remains as in all the other devices, the auto-discovery works. We could consider changing this API at a major version, but can't change it at a minor version release as it would break the usage people expect.

rmspeers commented on tools/zbkey:
```diff
@@ -111,7 +111,7 @@ DATA_KEY_SEQNUM = '\x03'
 while len(sys.argv) > 1:
     op = sys.argv.pop(1)
-    if op == '-f':
+    if op == '-f' or op == '-c':
```
zbkey should be moved to argparse, but we don't need to address that in this PR.

whoot replied:
I don't understand what you're trying to tell me.
I changed the -f parameter to -c to keep the parameters of the tools the same (in any other tool the channel is specified with -c, but not here). I kept the -f parameter for backwards compatibility.

rmspeers replied:
I'm just flagging that in some other update (doesn't have to be done here), we should be using argparse instead of sys.argv.pop. This code must have missed the prior update that the other tools got.

rmspeers commented on tools/zbreplay:
```diff
@@ -49,8 +49,12 @@ parser.add_argument('-s', '--sleep', action='store', type=float, default=1.0)
 
-if args.showdev:
-    show_dev()
+if args.showdev and not args.devstring:
```
See above note for this and other tools

whoot replied:
Same. Auto-Discovery does not work for me.

whoot commented:
So, you need anything else to add this request?

rmspeers commented:
I'm thinking about how to solve this overall, likely it requires changes to at least this branch, and possibly refactoring @AdamLaurie 's pages code. The decision mainly is to use things all-channel based or a mix of channels and pages.

I am currently leaning towards:
- move to only channels in the function signatures (remove pages, make this internal to firmware like SiLabs that uses them)
- make it the pattern that is_valid_channel can be overridden device driver code (else default applies)

@AdamLaurie  / @whoot  does the above sound reasonable?

The channel checking in zbdump will need to be removed, as will the Sewino specific interface enumeration behavior, but I can handle this when I update the branch for a change.

whoot commented:
Okay, so if I have understood that correctly, the channel checking should be in the device driver code?

Making it internal to firmware would be the best solution i think.

AdamLaurie commented:
The page thing, although I've only seen it so far with SiLabs, is, in fact, a "thing" in Europe. It is part of the Smart Metering standard so low level RF manufacturers may handle it differently, but users trying to adhere to the standards and/or test them, will be expecting to be able to specify page numbers. Maybe the simplest solution is to just allow both? Frequencies over 128 can be translated to frequency+page for that hardware that supports it, and others just ignore the page variable.

whoot committed:
Update dev_sewio.py

fixed typo which prevented jamming from working

whoot commented:
Yes, using both would be another way. But why dont we just define a different method for sewio which handles the frequency or why dont we just override the frequency method in dev_sewio?

whoot committed:
Update dev_sewio.py

fixing supported method/mode.
2.4 GHz does not support mode 4, 8, c, 10
all other frequencies dont support mode 20 or 40

whoot committed:
Create zbjammer

whoot committed:
Update setup.py

Add jamming tool for Sewio Open Sniffer

AdamLaurie commented:
It would be good if we supported the pages for hardware that is capable - the translation can be easily worked out if you look in kbutils.py - frequency() and is_valid_channel() both reference them. How are the various frequencies normally referenced on that hardware?

whoot commented:
The sewio is using the actual channel, modulation and RX values instead of pages.
I already mentioned that and referred to the documentation (https://www.sewio.net/open-sniffer/develop/http-rest-interface/).

In this case it would be useless to specify pages, if you cant use them,

Here are the specifications from the documentation:
![settings](https://user-images.githubusercontent.com/8066322/48476919-99a26180-e800-11e8-9b51-06be058dd8b0.png)
![frequencies](https://user-images.githubusercontent.com/8066322/48476922-9c9d5200-e800-11e8-8533-4eb847e76d8a.png)

AdamLaurie commented:
Right, so just leave page=0 and use channels as per the above as there is no direct mapping between the two. The operator will just have to figure it out based on what hardware he's using, so I guess we need to be very clear on frequency allocations in the documentation and/or comments.

whoot commented:
Okay, so I added the page=0 everywhere in the sewio code.
However, I dont know what you want me to do with the frequency() in kbutils.py

whoot commented:
Nothing new here?

whoot commented:
Alright.. 
I´m doing my best to add/update the Sewio branch based on the current development branch.
Will using Python3 only, since supporting Python2.7 will make no sense anymore.

rmspeers commented:
Thanks for bringing the Sewio branch up to date! Python3 is great.
